File: snowcap/api/rust/src/widget/signal.rs
use crate::signal::Signal;

/// Notifies that a redraw is needed.
#[derive(Signal, Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct RedrawNeeded;

/// Request the surface to close itself.
#[derive(Signal, Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct RequestClose;

/// Emits a message that will update widgets.
#[derive(Signal, Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Message<Msg>(pub Msg);

impl<Msg> Message<Msg> {
    /// Creates a new [`Message`].
    pub fn new(msg: Msg) -> Self {
        Self(msg)
    }

    /// Unwraps this message.
    pub fn into_inner(self) -> Msg {
        self.0
    }
}

impl<Msg> From<Msg> for Message<Msg> {
    fn from(value: Msg) -> Self {
        Self::new(value)
    }
}

/// Notifies that a widget closed.
#[derive(Signal, Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Closed;

File: snowcap/api/lua/snowcap/widget/signal.lua
-- This Source Code Form is subject to the terms of the Mozilla Public
-- License, v. 2.0. If a copy of the MPL was not distributed with this
-- file, You can obtain one at https://mozilla.org/MPL/2.0/.

---Signals emitted by widgets.
---
---@enum snowcap.widget.signal
return {
    ---Notifies that a redraw is needed.
    redraw_needed = "widget::redraw_needed",
    ---Emits a message that will update widgets.
    send_message = "widget::send_message",
    ---Notifies that a widget closed.
    closed = "widget::closed",

    ---Update widgets' internal state.
    operation = "widget::operation",
}
